keps/sig-api-machinery/4933-cacher-label-index/README.md

```diff
@@ -0,0 +1,880 @@
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Since this feature is relatively large, the entire feature will be controlled by a feature gate `CacherLabelIndex` in the early stage.
+
+Considering that daemonset and spark applications coexist in a cluster, and they watch based on `spec.nodeName` and `applicationId` label respectively, we will support both usage at the same time.
+
+In reality, a single filter condition, such as nodeName or applicationId, can already accurately match cacheWatcher, this feature still considers supporting only one triggerIndex for a watch request. At the same time, since the following prerequisites guarantee that OR operations are not supported, filtering events with a single triggerIndex can ensure correctness
+- Label selectors and field selectors in a watch request are AND operations.
+- Multiple requirements in label selector are also AND operations.
+- Through a [`RequiresExactMatch`](https://github.com/kubernetes/kubernetes/blob/v1.31.2/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L574) check, a set-based label selector or field selector that supports OR operations on multiple values is considered not matching trigger indexes in cacher.
+
+### Commandline changes
+
+Considering that label names may be different in different scenarios, such as app label name in Flink and spark-app-selector label name in Spark, this feature will add a commandline parameter to the apiserver, allowing different label names to be configured as the trigger index of the apiserver cacher.
+
+```
+--index-labels. Index labels settings for some resources (pods, nodes, etc.), comma separated. The individual setting format: resource[.group]#labelName1;labelName2, where resource is lowercase plural (no version), group is omitted for resources of apiVersion v1 (the legacy core API) and included for others, and labelName is a string. It takes effect when watch-cache and feature gate CacherLabelIndex are enabled. In addition to built-in indexes such as pod.spec.nodeName, apiserver cacher will create additional indexes for these labels to speed up list and watch requests.
+
+# e.g. for spark
+--index-labels=pods#spark-app-selector
+# e.g. for spark and flink
+--index-labels=pods#spark-app-selector;app
+```
+
+### Changes of register trigger index and cache index to resource storage
```

Comment:
I would like to take this opportunity to do one more thing - not only optimize watchers, but also be able to optimize list requests.
We already do support multiple indexes for store:
https://github.com/kubernetes/kubernetes/blob/5d9144ca2ef9ffa0fe88e011cdecf8182eaad22d/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L566-L574

but let's take the opportunity of this KEP to also:
- register those indexes (use the same labels that you pass via commandline flags - just pass them both as indexers for watches as well as indexers for store)
- implement that TODO to choose the minimum of the values there

Reply:
Thanks, I added "improve list performance by using multiple label indexes and field indexes" to record this.